massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1074 @@
+            MipState::LockedIn(_) => MipStateTypeId::LockedIn,
+            MipState::Active(_) => MipStateTypeId::Active,
+            MipState::Failed(_) => MipStateTypeId::Failed,
+        }
+    }
+}
+
+/// A message to update the `MipState`
+#[derive(Clone, Debug)]
+pub struct Advance {
+    /// from VersioningInfo.start
+    pub start_timestamp: MassaTime,
+    /// from VersioningInfo.timeout
+    pub timeout: MassaTime,
+    /// % of past blocks with this version
+    pub threshold: Amount,
+    /// Current time (timestamp)
+    pub now: MassaTime,
+}
+
+// Need Ord / PartialOrd so it is properly sorted in BTreeMap
+
+impl Ord for Advance {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.now).cmp(&other.now)
+    }
+}
+
+impl PartialOrd for Advance {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Advance {
+    fn eq(&self, other: &Self) -> bool {
+        self.start_timestamp == other.start_timestamp
+            && self.timeout == other.timeout
+            && self.threshold == other.threshold
+            && self.now == other.now
+    }
+}
+
+impl Eq for Advance {}
+
+transitions!(MipState,
+    [
+        (Defined, Advance) => [Defined, Started, Failed],
+        (Started, Advance) => [Started, LockedIn, Failed],
+        (LockedIn, Advance) => [LockedIn, Active],
+        (Active, Advance) => Active,
+        (Failed, Advance) => Failed
+    ]
+);
+
+impl Defined {
+    /// Update state from state Defined
+    pub fn on_advance(self, input: Advance) -> MipState {
+        match input.now {
+            n if n > input.timeout => MipState::failed(),
+            n if n > input.start_timestamp => MipState::started(Amount::zero()),
+            _ => MipState::Defined(Defined {}),
+        }
+    }
+}
+
+impl Started {
+    /// Update state from state Started
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            return MipState::failed();
+        }
+
+        if input.threshold >= VERSIONING_THRESHOLD_TRANSITION_ACCEPTED {
+            MipState::locked_in()
+        } else {
+            MipState::started(input.threshold)
+        }
+    }
+}
+
+impl LockedIn {
+    /// Update state from state LockedIn ...
+    pub fn on_advance(self, input: Advance) -> MipState {
+        if input.now > input.timeout {
+            MipState::active()
+        } else {
+            MipState::locked_in()
+        }
+    }
+}
+
+impl Active {
+    /// Update state (will always stay in state Active)
+    pub fn on_advance(self, _input: Advance) -> Active {
+        Active {}
+    }
+}
+
+impl Failed {
+    /// Update state (will always stay in state Failed)
+    pub fn on_advance(self, _input: Advance) -> Failed {
+        Failed {}
+    }
+}
+
+/// Wrapper of MipState (in order to keep state history)
+#[derive(Debug, Clone, PartialEq)]
+pub struct MipStateHistory {
```

Comment:
I don't find this structure very clear. And what is his purpose. Why we are not using a MipState directly ?

EDIT: Reading the rest of the code this allow to have the history of each state of the version. Is that useful ? Would it be send through the bootstrap ?

Reply:
The history is useful to know the state of a proposal for a given timestamp. For example, when we receive a block header in the protocol-worker with a timestamp `block_timestamp < now()`, we check the history to see if we need to keep or reject the block header.
We need to know if the "active version" of the block is a version that was in active state at `block_timestamp`, and not at `now()`.

Reply:
Ok why we don't save only the list of the current active version and the timestamp when they became active ?

Reply:
I've added some doc in lib.rs about when we used the history